File: 백준/Gold/2252. 줄 세우기/줄 세우기.cc
#include <bits/stdc++.h>

using namespace std;

#define MAX 32000

int N, M;
vector<int> graph[MAX + 1];
vector<int> degree;
queue<int> q;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);

    cin >> N >> M;

    degree.assign(N + 1, 0);

    int a, b;
    for (int i = 0; i < M; i++) {
        cin >> a >> b;
        graph[a].emplace_back(b);
        degree[b]++;
    }

    for (int i = 1; i <= N; i++) {
        if (degree[i] == 0) q.push(i);
    }

    while (!q.empty()) {
        int cur = q.front();
        q.pop();
        printf("%d ", cur);
        for (int next: graph[cur]) {
            if (--degree[next] == 0) q.push(next);
        }
    }
    
    return 0;
}

File: 백준/Gold/2252. 줄 세우기/줄 세우기.js
const readline = require('readline');
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});
const input = [];

rl.on('line', (line) => {
  input.push(line);
}).on('close', () => {
  const [N] = input[0].split(' ').map(Number);
  const edges = input.slice(1).map((v) => v.split(' ').map(Number));

  solution(N, edges);
});

const solution = (N, edges) => {
  const degree = Array(N + 1).fill(0);
  const graph = Array.from(Array(N + 1), () => Array());
  const queue = new Queue();
  const result = [];

  for (const [a, b] of edges) {
    graph[a].push(b);
    degree[b] += 1;
  }

  for (let i = 1; i <= N; i++) {
    if (degree[i] === 0) queue.push(i);
  }

  while (!queue.isEmpty()) {
    const current = queue.pop();

    result.push(current);

    for (const next of graph[current]) {
      degree[next] -= 1;
      if (degree[next] === 0) queue.push(next);
    }
  }

  console.log(result.join(' '));
};

class Queue {
  constructor() {
    this.queue = [];
    this.front = 0;
    this.rear = 0;
  }

  push(item) {
    this.queue[this.rear++] = item;
  }

  pop() {
    const value = this.queue[this.front];
    delete this.queue[this.front];
    this.front += 1;
    return value;
  }

  isEmpty() {
    return this.front === this.rear;
  }
}
